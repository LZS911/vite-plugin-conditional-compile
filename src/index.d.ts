import { PluginOption } from 'vite';

export type Options = {
  isDebug?: boolean;
  changeSource?: (source: string) => string;
  expand?: {
    [key: string]: boolean | undefined;
  };
};

declare function vitePluginConditionalCompile(opts?: Options): PluginOption[];

export default vitePluginConditionalCompile;
