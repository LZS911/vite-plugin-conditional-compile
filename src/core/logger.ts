import { createLogger } from 'vite';

export const logger = createLogger(undefined, { prefix: 'vite-plugin-conditional-compile' });
