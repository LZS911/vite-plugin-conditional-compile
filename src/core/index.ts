import { parseSync, traverse } from '@babel/core';
import { createFilter } from 'vite';
import { Options } from '../index.d';
import { logger } from './logger';
import {
  BinaryExpression,
  Identifier,
  LogicalExpression,
  NumericLiteral,
  UnaryExpression,
  ExpressionStatement,
} from '@babel/types';

type ConditionalSupportType =
  | BinaryExpression
  | Identifier
  | LogicalExpression
  | NumericLiteral
  | UnaryExpression;

class Context {
  private ctx: Options = {} as Options;
  private code: string = '';
  private id: string = '';

  constructor(userOptions: Options) {
    this.ctx = userOptions;
  }

  public set_env(env: Record<string, any>) {
    this.ctx.env = { ...this.ctx.env, ...env };
  }

  public resolve_options(userOptions: Options) {
    return {
      include: ['**/*'],
      exclude: [/[\\/]node_modules[\\/]/, /[\\/]\.git[\\/]/],
      ...userOptions,
    };
  }

  public evaluate_ast(node: ConditionalSupportType): boolean {
    console.log(node);
    try {
      if (node.type === 'BinaryExpression') {
        const { left, operator, right } = node;
        if (left.type === 'Identifier' && ['==', '!=='].includes(operator)) {
          if (right.type === 'Identifier') {
            if (operator === '==') {
              return this.ctx.env?.[left.name] == right.name;
            }
            return this.ctx.env?.[left.name] != right.name;
          }

          if (right.type === 'NumericLiteral') {
            if (operator === '==') {
              return this.ctx.env?.[left.name] == right.value;
            }
            return this.ctx.env?.[left.name] != right.value;
          }
        } else {
          logger.error(
            `evaluate ast error: This expression is not supported: ${left.type}${operator}${right.type}`,
          );
          return false;
        }
      }

      if (node.type === 'Identifier') {
        return this.ctx.env?.[node.name] === true;
      }

      if (node.type === 'NumericLiteral') {
        return this.ctx.env?.[node.value] === true;
      }

      if (node.type === 'UnaryExpression') {
        if (node.operator === '!') {
          if (node.argument.type === 'Identifier') {
            return this.ctx.env?.[node.argument.name] === false;
          }
          if (node.argument.type === 'UnaryExpression') {
            return this.evaluate_ast(node.argument);
          }
          logger.error(
            `evaluate ast error: This expression is not supported: ${node.type}:${node.operator}`,
          );
          return false;
        } else {
          logger.error(
            `evaluate ast error: This expression is not supported: ${node.type}:${node.operator}`,
          );
          return false;
        }
      }

      if (node.type === 'LogicalExpression') {
        const { left, operator, right } = node;

        if (operator === '&&') {
          return (
            this.evaluate_ast(left as ConditionalSupportType) &&
            this.evaluate_ast(right as ConditionalSupportType)
          );
        }
        if (operator === '||') {
          return (
            this.evaluate_ast(left as ConditionalSupportType) ||
            this.evaluate_ast(right as ConditionalSupportType)
          );
        }
        if (operator === '??') {
          return (
            this.evaluate_ast(left as ConditionalSupportType) ??
            this.evaluate_ast(right as ConditionalSupportType)
          );
        }
      }

      logger.error(`evaluate ast error: This expression is not supported: ${node.type}`);
      return false;
    } catch (error) {
      logger.error(`evaluate ast error: ${error}`);
      return false;
    }
  }

  public resolve_conditional(conditional: string): boolean {
    let res = false;
    try {
      const ast = parseSync(conditional.replace(/([^=!])=([^=])/g, '$1==$2'));
      if (!ast) {
        return false;
      }

      traverse(ast, {
        Program: ({ node }) => {
          res = this.evaluate_ast(
            (node.body[0] as ExpressionStatement).expression as ConditionalSupportType,
          );
        },
      });
    } catch (error) {
      logger.error(`${this.id}: babel parse error: ${error}`);
      return false;
    }
    return res;
  }

  public resolve_comment() {
    const pattern_reg =
      /(?:\/\/|\{?\/\*)\s*#\s*if\s\[([\w!=&|()'"\s]+)\](?:\s*\*\/\}?)?([\s\S]+?)(?:\/\/|\{?\/\*)\s*#\s*endif(?:\s*\*\/\}?)?/g;

    return this.code.replace(pattern_reg, (matchCode) => {
      const match_group_reg =
        /.*(#\s*(?:if|elif|else|endif))\s?(?:\[([\w !=&|()'"]*)\])*(?:\s*\*\/\s*\})?/g;
      const codeBlocks = matchCode.split(match_group_reg).filter((v) => v !== '');
      while (codeBlocks.length) {
        const [flag, conditional, block] = codeBlocks.splice(0, 3);
        const real_flag = flag.replace(/\s/g, '');
        if (real_flag === '#if' || real_flag === '#elif') {
          if (this.resolve_conditional(conditional)) {
            return block;
          }
        } else if (real_flag === '#else') {
          return block;
        }
      }

      return '';
    });
  }

  public deprecated_transform(code: string) {
    if (!('isDebug' in this.ctx)) {
      this.ctx.isDebug = !!this.ctx.env?.DEV;
    }
    try {
      const js = this.ctx?.changeSource?.(code) ?? code;
      const replaceMatched = (): string => {
        const REG = /\/\*\s*IF(DEBUG|TRUE_\w+)(?:\s*\*\/)?([\s\S]+?)(?:\/\*\s*)?FI\1\s*\*\//g;
        return js.replace(REG, (_, $1, $2) => {
          let isKeep = false;
          if ($1 === 'DEBUG') {
            isKeep = this.ctx.isDebug!;
          } else {
            isKeep = !!this.ctx.expand?.[$1.slice(5)];
          }

          return isKeep ? $2 : '';
        });
      };

      return replaceMatched();
    } catch (error) {
      throw error;
    }
  }

  public transform(code: string, id: string) {
    this.code = code;
    this.id = id;

    const options = this.resolve_options({ include: this.ctx.include, exclude: this.ctx.exclude });
    const filter = createFilter(options.include, options.exclude);
    if (!filter(id)) {
      return this.deprecated_transform(code);
    }

    return this.deprecated_transform(this.resolve_comment());
  }
}

export const create_context = (userOptions: Options) => {
  return new Context(userOptions);
};
