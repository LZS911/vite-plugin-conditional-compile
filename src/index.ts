import { loadEnv, Plugin } from 'vite';
import { create_context } from './core';
import { Options } from './index.d';

const vitePluginConditionalCompile: (userOptions: Options) => Plugin = (userOptions = {}) => {
  const ctx = create_context(userOptions);
  return {
    name: 'vite-plugin-conditional-compile',
    enforce: 'pre',
    configResolved(config) {
      ctx.set_env({
        ...loadEnv(config.mode ?? process.env.NODE_ENV, process.cwd(), ''),
        ...config.env,
      });
    },
    transform(code, id) {
      return ctx.transform(code, id);
    },
  };
};

export default vitePluginConditionalCompile;
