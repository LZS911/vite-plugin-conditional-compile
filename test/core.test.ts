import { create_context } from '../src/core';
import {
  mock_condition_txt,
  mock_condition_txt_2,
  mock_condition_txt_3,
  mock_condition_txt_4,
  mock_condition_txt_5,
  mock_condition_txt_6,
  mock_condition_txt_7,
  mock_condition_txt_8,
  mock_deprecated_txt,
  mock_match_txt_1,
  mock_match_txt_2,
  mock_match_txt_3,
  mock_match_txt_4,
  mock_match_txt_5,
  mock_txt_1,
} from './code/test_txt';

const remove_space = (str: string) => str.replace(/\s/g, '');

describe('test', () => {
  const ctx = create_context({
    include: /\/src\/.*\.tsx*/,
    expand: {
      isDebug: true,
      isCond: true,
      isProd: false,
    },
  });
  beforeAll(() => {
    ctx.set_env({
      dev: true,
      feature: 'cond',
      prod: false,
      prod_version: 'ce',
    });
  });

  it('test deprecated transform', () => {
    expect(remove_space(ctx.transform(mock_deprecated_txt, '/src/a.js'))).toBe(
      remove_space(`
  if true dev  block
  {}
  {if true cond  block}
  {}{}
        `),
    );
  });

  it('test filter', () => {
    expect(remove_space(ctx.transform(mock_txt_1, '/src/a.js'))).toBe(remove_space(mock_txt_1));

    expect(remove_space(ctx.transform(mock_txt_1, '/src/a.ts'))).toBe(
      remove_space(`
    AAAAAA
    3ufhdsajklnvc832h
    fdsahfjk023dfjs
    fdslkfjfklda
    BBBBBBB`),
    );
  });

  it('test match', () => {
    expect(ctx.pattern_reg.test(mock_match_txt_1)).toBeTruthy();
    expect(ctx.pattern_reg.test(mock_match_txt_2)).toBeFalsy();
    expect(ctx.pattern_reg.test(mock_match_txt_3)).toBeTruthy();
    expect(ctx.pattern_reg.test(mock_match_txt_4)).toBeFalsy();
    expect(ctx.pattern_reg.test(mock_match_txt_5)).toBeTruthy();
  });

  it('test conditional', () => {
    expect(remove_space(ctx.transform(mock_condition_txt, '/src/a.ts'))).toBe(
      remove_space(
        `
      AAAAAA
      else block
      BBBBBBB
      `,
      ),
    );

    expect(remove_space(ctx.transform(mock_condition_txt_2, '/src/a.ts'))).toBe(
      remove_space(
        `
      AAAAAA
      if block
      BBBBBBB
      `,
      ),
    );

    expect(remove_space(ctx.transform(mock_condition_txt_3, '/src/a.ts'))).toBe(
      remove_space(
        `
      AAAAAA
      elif block
      BBBBBBB
      `,
      ),
    );

    expect(remove_space(ctx.transform(mock_condition_txt_4, '/src/a.ts'))).toBe(
      remove_space(
        `
      AAAAAA
      elif block
      BBBBBBB
      `,
      ),
    );

    expect(remove_space(ctx.transform(mock_condition_txt_5, '/src/a.ts'))).toBe(
      remove_space(
        `
      AAAAAA
      else block
      BBBBBBB
      `,
      ),
    );

    expect(remove_space(ctx.transform(mock_condition_txt_6, '/src/a.ts'))).toBe(
      remove_space(
        `
      AAAAAA
      else block
      BBBBBBB
      `,
      ),
    );

    expect(remove_space(ctx.transform(mock_condition_txt_7, '/src/a.ts'))).toBe(
      remove_space(`
      AAAAAA
      else block
      BBBBBBB
    `),
    );

    expect(remove_space(ctx.transform(mock_condition_txt_8, '/src/a.ts'))).toBe(
      remove_space(`
      AAAAAA
      <span>elif block</span>
      BBBBBBB
    `),
    );
  });
});
