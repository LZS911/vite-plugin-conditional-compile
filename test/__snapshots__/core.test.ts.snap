// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test sourceMaps dev mode 1`] = `
{
  "file": undefined,
  "ignoreList": [],
  "mappings": "AAGA,OAAOA,IAAI,MAAM,MAAM",
  "names": [
    "path",
  ],
  "sourceRoot": undefined,
  "sources": [
    "/src/a.js",
  ],
  "sourcesContent": [
    "


import path from 'path'

",
  ],
  "version": 3,
}
`;
